### ISSUE TITLE
Create interpolation function

### ISSUE THREAD

**Issue by jamenkaye** [UBCSailbot/local_pathfinding#28]
### Purpose

We want to be able to calculate how long the boat will take to travel a given path in known wind conditions. To do this, we need to know the boat speed, which depends on the "[point of sail](https://ubcsailbot.github.io/docs/main/reference/sailing/points_of_sail/)" and the windspeed.

### Description

I've come up with a table of estimates for boat speed, based on the windspeed and sailing angle. This task is to come up with a function to interpolate this data:

Windspeed | 0 deg | 20 deg | 30 deg | 45 deg | 90 deg | 135 deg | 180 deg
-- | -- | -- | -- | -- | -- | -- | --
0      | 0 | 0 | 0 | 0 | 0 | 0 | 0
9.3   | 0 | 0 | 0.4 | 1.1 | 3.2 | 3.7 | 2.8
18.5 | 0 | 0.3 | 1.9 | 3.7 | 9.3 | 13.0 | 9.2
27.8 | 0 | 0.9 | 3.7 | 7.4 | 14.8 | 18.5 | 13.0
37.0 | 0 | 1.3 | 5.6 | 9.3 | 18.5 | 24.1 | 18.5

The exact values in the table _**or the dimensions of the table might get changed later**_ so make sure your implementation is flexible enough to allow this easily.

The function arguments will be:
- [True Windspeed](https://ubcsailbot.github.io/docs/main/reference/sailing/miscellaneous/#:~:text=angle%20conventions%20are.-,True%2C%20Apparent%2C%20and%20Boat%20Wind,-%C2%B6) in km/h (Make sure function header states that it's True Wind)
- Sailing angle in degrees (Eg, 0 degrees means the boat is in Irons, 45 degrees means the boat is on Close Hauled, 180 degrees means the boat is on a Run)

The function return will be:
- Boat speed in km/h

There are tons of different mathematical approaches to interpolation. I would suggest starting with something like the simple approach shown below, but you are free to look into more complicated interpolations if you like :)

1. Suppose function inputs are 60 degrees, 12 knots.
2. Lets look at the 10-knot row of the table first. $2 \cdot \frac{90-60}{90-45} + 5 \cdot \frac{60-45}{90-45} = 3$ so our first intermediate value is 3 knots.
3. Now lets look at the 15 knot row of the table. $4 \cdot \frac{90-60}{90-45} + 8 \cdot \frac{60-45}{90-45} = 5.33$, which is our second intermediate value.
4. Last, we'll interpolate between the 10 and 15-knot rows, using the "intermediate" values from steps 2 and 3: $3 \cdot \frac{15-12}{15-10} + 5.33 \cdot \frac{12-10}{15-10} = 3.93$. Using this approach, our final answer for the boat speed is 3.93 knots.

Make sure to include a short function header to explain the purpose of the function, and what the arguments and return value are. Also, think about what values for the inputs might be permissible, and note any restrictions in the function description.

Write your interpolation function in the path_objective.py file.

If you feel up for it, it would also be nice to try to visualize the data with a 3D surface plot. [Plotly](https://plotly.com/python/3d-charts/) has some nice looking options, for example, [this one](https://plotly.com/python/3d-surface-plots/). Create a folder called `research` in the `local_pathfinding` repository and make a new file in this folder to write your plotting code. Note: if you ever need to show off your programming experience in a visual kind of way, a colorful 3D plot would probably look amazing 👀

### Summary
1. Write a linear interpolation function.
2. (bonus) Plot the data as a 3D surface to evaluate the interpolation method (and also help me revise my estimates of the boat speed)
3. (bonus) Research and write a more complicated interpolation function. If you do this, do not delete your linear function; save them both to compare against each other.

### Resources

- https://ubcsailbot.github.io/docs/main/reference/sailing/points_of_sail/
- https://ubcsailbot.github.io/docs/main/reference/sailing/miscellaneous/#:~:text=A-,Relative%20Bearing,-is%20a%20bearing You could use the term "Relative Bearing" to help make a thorough description of what the wind angle is
- Plotly all 3d plots: https://plotly.com/python/3d-charts/
- https://matplotlib.org/stable/gallery/mplot3d/index.html

**Comment by jamenkaye** [comment]
Todo before this issue is ready:
- [x] Get @FireBoyAJ24 to look over, and determine where this function should be written
- [x] Get @patrick-5546 to confirm plotting code for a "visual test" can go in a new file in the `test` folder (Answer: no, make `research` folder)

**Comment by jamenkaye** [comment]
Discussed this issue today during work party. 
Issue: Manual calculations and the function output disagree by ~0.1 km/h boatspeed. This isn't a huge issue for itself, but I'm concerned it could indicate some other issue. We should either understand this or fix this before the interpolation function is finalized.

Next step: Write some tests to try to figure out what the issue is.
- You can just write the tests in a new cell at the bottom. (Remember you need to run the cell where the function is defined before you can run you tests)
- We'd like to have some tests go in the `local_pathfinding/test` folder for interpolation anyway, so you may as well write your tests now. Check out what tests are already there, and try to match the same style.
- Make sure to test special cases, eg: input value is exactly a point in the table, input is outside the range covered by the table. 

You can also try using print statements in the middle of your code to isolate the issue.

**Comment by jamenkaye** [comment]
How to handle if the input is exactly on one row of the table?

Here's one possible approach:
- If the angle is 0 or the windspeed is 0, return that the boat speed is 0.
- Otherwise, start checking through the list of windspeed values in the table to see when the $\text{input windspeed} \leq \text{table windspeed}$. We then interpolate between the $\text{table windspeed}$ row that we just got to, and the previous row.
     - As a standard example, the input windspeed is 12 km/h. Then the first time $\text{input windspeed} \leq \text{table windspeed}$ is when $\text{table windspeed} = 18.5$. We will interpolate between the 9.3 row and the 18.5 row.
     - As an "edge case" example, the input windspeed is exactly 18.5 km/h. Then the first time $\text{input windspeed} \leq \text{table windspeed}$ is when $\text{table windspeed} = 18.5$. We will interpolate between the 9.3 row and the 18.5 row. BUT, the values in the 9.3 will end up with a "weight" of 0, because we are right on top of the 18.5 row.

You should be able to use this same kind of approach for the angles as well.

**Comment by jamenkaye** [comment]
What if the input values are outside the table?
- The angle cannot be outside of [0, 180] (may need to take absolute value tho)
- If the windspeed is higher than the maximum, just assume that its equal to the maximum and interpolate between two values in the maximum row.
     - This is acceptable because we care most about how fast different angles are relative to each other, at the same windspeed.

When you're ready to push code:
- Check out the section on the [docs site](https://ubcsailbot.github.io/docs/main/reference/github/workflow/overview/) about pushing code and PRs
- Ask Jay where the function should go in the local pathfinding repository
- Add your tests to the test folder in local pathfinding
- Make sure to ctrl-s to manually save your code, which runs the linters and removes trailing whitespace and stuff
- Then we can do the code review process in the local pathfinding repo

**Comment by jamenkaye** [comment]
Remaining TODOs:
- [ ] Create a new branch from Jay's branch of local pathfinding: (see Patrick's comment below)
- [ ] Copy and paste the contents of the function from the notebooks repository. The destination is line 465 in `objectives.py`
- [ ] Commit and push to your new speed-interpolation branch. See the [docs site](https://ubcsailbot.github.io/docs/latest/reference/github/workflow/branches/#tracking-and-committing-changes) for details
- [ ] Create a PR to merge your speed-interpolation branch into Jay's branch `Skeleton-Speed-Objective-Akshanjay`

**Comment by patrick-5546** [comment]
> Create a PR to merge your speed-interpolation branch into Jay's branch `Skeleton-Speed-Objective-Akshanjay`

Remember to branch off from this branch to avoid headaches: 

```
git switch main
git pull
git switch Skeleton-Speed-Objective-Akshanjay
git switch -c justin-prasad/speed-objective
```

**Comment by FireBoyAJ24** [comment]
Resolved by #82
